substrate/client/network/src/peer_info.rs

```diff
@@ -378,23 +404,41 @@ impl NetworkBehaviour for PeerInfoBehaviour {
 				self.ping.on_swarm_event(FromSwarm::ListenerError(e));
 				self.identify.on_swarm_event(FromSwarm::ListenerError(e));
 			},
-			FromSwarm::ExpiredExternalAddr(e) => {
-				self.ping.on_swarm_event(FromSwarm::ExpiredExternalAddr(e));
-				self.identify.on_swarm_event(FromSwarm::ExpiredExternalAddr(e));
-			},
 			FromSwarm::NewListener(e) => {
 				self.ping.on_swarm_event(FromSwarm::NewListener(e));
 				self.identify.on_swarm_event(FromSwarm::NewListener(e));
 			},
+			FromSwarm::NewListenAddr(e) => {
+				self.ping.on_swarm_event(FromSwarm::NewListenAddr(e));
+				// Only notify [`Identify`] if we do not have the explicit list of advertised
+				// addresses.
+				if self.public_addresses_only.is_none() {
+					self.identify.on_swarm_event(FromSwarm::NewListenAddr(e));
+				}
+			},
 			FromSwarm::ExpiredListenAddr(e) => {
 				self.ping.on_swarm_event(FromSwarm::ExpiredListenAddr(e));
-				self.identify.on_swarm_event(FromSwarm::ExpiredListenAddr(e));
-				self.external_addresses.remove(e.addr);
+				// See `FromSwarm::NewListenAddr` for why we do not always notify [`Identify`].
+				if self.public_addresses_only.is_none() {
+					self.identify.on_swarm_event(FromSwarm::ExpiredListenAddr(e));
+				}
 			},
 			FromSwarm::NewExternalAddr(e) => {
 				self.ping.on_swarm_event(FromSwarm::NewExternalAddr(e));
-				self.identify.on_swarm_event(FromSwarm::NewExternalAddr(e));
-				self.external_addresses.add(e.addr.clone());
+				// Only notify [`Identify`] and add an external address if we do not have the
+				// explicit list of advertised addresses.
+				if self.public_addresses_only.is_none() {
+					self.identify.on_swarm_event(FromSwarm::NewExternalAddr(e));
+					self.external_addresses.add(e.addr.clone());
+				}
+			},
+			FromSwarm::ExpiredExternalAddr(e) => {
```

Comment:
What would happen if the external address that expired was part of the initial `public_addresses_only`?

We'd keep advertising the same addresses, although we'd know that they expired?

Reply:
Explicitly set external addresses never expire, as they are added with [`AddressScore::Infinite`](https://docs.rs/libp2p/0.51.3/libp2p/swarm/enum.AddressScore.html):
https://github.com/paritytech/polkadot-sdk/blob/4987d7982461e2e5ffe219cdf71ec697284cea7c/substrate/client/network/src/service.rs#L500-L506

Reply:
Cool! Thanks for the info! 🙏